### PR TITLE
Add redux to app

### PR DESCRIPTION
**Pull request by XeniaVi**

### CODE REVIEW

File: src/api/todoApi.js [hunk 1]
```diff
@@ -18,8 +18,8 @@ export const deleteTodoFromDB = async (id) => {
   return axios.delete(`${config.API_URL}/${id}`, id);
 };
 
-export const updateTodoInDB = async (id, post) => {
-  const response = await axios.put(`${config.API_URL}/${id}`, post);
+export const updateTodoInDB = async (id, updatedTodo) => {
```
Comment (slimShady88) [inline src/api/todoApi.js]: Let's remove ToDB, InDB or FromDB from the names of such functions

File: src/reducers/TodoReducer.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+      todos = state.todos.map((item) =>
+        item.id === id
+          ? {
+              ...item,
+              ...updatedTodo,
+            }
+          : item
+      );
+
+      return { ...state, todos: todos };
```
Comment (slimShady88) [inline src/reducers/TodoReducer.js]: There is no need to duplicate `{todos: todos}`, it can be just `{todos}`

File: src/reducers/TodoReducer.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+      todos = state.todos.map((item) =>
+        item.completed !== payload
+          ? {
+              ...item,
+              completed: payload,
+            }
+          : item
+      );
+
+      return { ...state, todos: todos };
```
Comment (slimShady88) [inline src/reducers/TodoReducer.js]: Here the same

File: src/asyncActions/addTodo.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+} from "../actions";
+import { addTodoToDB } from "../api/todoApi";
+
+export const addTodo = (value) => {
+  return async (dispatch) => {
+    try {
+      const response = await addTodoToDB({
+        value: value,
+        completed: false,
+        timestamp: Date.now(),
```
Comment (slimShady88) [inline src/asyncActions/addTodo.js]: createdAt is best practice

File: src/asyncActions/fetchTodos.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
+import { getTodosAction } from "../actions";
+import {
```
Comment (slimShady88) [inline src/asyncActions/fetchTodos.js]: Duplicate import

File: src/asyncActions/fetchTodos.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  setErrorAction,
+} from "../actions";
+import { getTodos } from "../api/todoApi";
+import { LIMIT } from "../constants";
+
+export const fetchTodos = (offset, completed) => {
+  return async (dispatch) => {
+    try {
+      const response = await getTodos(LIMIT, offset, completed);
+      const items = response.todos;
```
Comment (slimShady88) [inline src/asyncActions/fetchTodos.js]: It can be named 'todos' and then destructurization can be used

File: src/asyncActions/updateTodos.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
+import { updateTodosAction, setErrorAction } from "../actions";
+import { setCompletedAllAction } from "../actions";
```
Comment (slimShady88) [inline src/asyncActions/updateTodos.js]: Duplicate import

File: src/api/todoApi.js (hunk 1 above)
Comment (igirin) [inline src/api/todoApi.js:21]: Lets call it just updateTodo

File: src/components/App.js [hunk 8]
```diff
@@ -12,294 +24,103 @@ import {
...
-
-  const handleChangeInputCheckbox = () => {
-    setCompletedAll(!completedAll);
-  };
-
-  const handleChangeItem = (id, completed) => {
-    changeStatus(id, completed);
+    const ids = items
+      .filter((item) => item.completed)
+      .map((item) => {
```
Comment (slimShady88) [inline src/components/App.js]: This .map function can be written in one line

File: src/constants/index.js [hunk 9]
```diff
@@ -0,0 +1,15 @@
+export const GET_TODOS = "GET_TODOS";
```
Comment (nortoza) [inline src/constants/index.js]: Lets make these constants as const object with these fields

File: src/constants/index.js [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+export const SET_COMPLETED = "SET_COMPLETED";
+export const SET_COMPLETED_ALL = "SET_COMPLETED_ALL";
+export const SET_PAGE = "SET_PAGE";
+export const SET_OFFSET = "SET_OFFSET";
+export const SET_COUNT = "SET_COUNT";
+export const SET_ERROR = "SET_ERROR";
+export const SET_FILTER = "SET_FILTER";
+
+export const LIMIT = 5;
```
Comment (nortoza) [inline src/constants/index.js]: Lets call it more specified

File: src/components/App.js [hunk 11]
```diff
@@ -12,294 +24,103 @@ import {
...
+    dispatch(setErrorAction(""));
+    dispatch(setCountAction(items.filter((item) => !item.completed).length));
   }, [items]);
 
-  useEffect(() => {
-    fetchTodos(limit, offset, completed);
-  }, [page, completed]);
-
   return (
     <div>
```
Comment (slimShady88) [inline src/components/App.js]: Useless div

File: src/asyncActions/deleteTodo.js [hunk 12]
```diff
@@ -0,0 +1,16 @@
+import { fetchTodos } from "./fetchTodos";
+import { deleteTodoFromDB } from "../api/todoApi";
+import { setErrorAction } from "../actions";
+
```
Comment (nortoza) [inline src/asyncActions/deleteTodo.js]: Lets move asyncActions to one file "asyncActions.js"

File: src/components/App.js [hunk 13]
```diff
@@ -12,294 +24,103 @@ import {
...
-          />
-          <Pagination
-            page={page}
-            switchPages={switchPages}
-            totalCount={totalCount}
-            limit={limit}
-          />
+          <InputTask />
+          <Pagination />
           {items.length ? (
```
Comment (slimShady88) [inline src/components/App.js]: Let's use `items.length && (...some)` instead `items.length ? (...some) : ""`

File: src/asyncActions/fetchTodos.js [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+
+      dispatch(getTodosAction(response));
+      dispatch(
+        setCompletedAllAction(
+          items.length === items.filter((item) => item.completed).length
```
Comment (nortoza) [inline src/asyncActions/fetchTodos.js]: Lets move filter operations with getting length to consts

File: src/components/App.js [hunk 15]
```diff
@@ -12,294 +24,103 @@ import {
...
-  const handleChangeInputCheckbox = () => {
-    setCompletedAll(!completedAll);
-  };
-
-  const handleChangeItem = (id, completed) => {
-    changeStatus(id, completed);
+    const ids = items
+      .filter((item) => item.completed)
+      .map((item) => {
+        return item.id;
```
Comment (nortoza) [inline src/components/App.js]: We can remove return and brackets

File: src/components/Pagination.js [hunk 16]
```diff
@@ -8,22 +9,19 @@ import {
...
-  };
-
-  useEffect(() => countPages(), [totalCount]);
+  const pages = useSelector((state) => state.todos.pages);
+  const page = useSelector((state) => state.status.page);
 
   const getPageList = () => {
     const list = [];
-    const length = pagesCount >= 3 ? 3 : pagesCount;
+    const length = pages >= 3 ? 3 : pages;
```
Comment (nortoza) [inline src/components/Pagination.js]: Why renamed from pagesCount -> pages?

File: src/components/InputTask.js [hunk 17]
```diff
@@ -1,23 +1,59 @@
...
+  };
+
+  const handleChangeInputCheckbox = () => {
+    dispatch(setCompletedAllAction(completedAll));
+  };
+
   return (
     <InputWrapper>
-      {length ? (
+      {items.length ? (
```
Comment (slimShady88) [inline src/components/InputTask.js]: The same, just `items.length && ...`

File: src/components/Pagination.js [hunk 18]
```diff
@@ -8,22 +9,19 @@ import {
...
+function Pagination() {
+  const dispatch = useDispatch();
 
-  const countPages = () => {
-    setPagesCount(Math.ceil(totalCount / limit));
-  };
-
-  useEffect(() => countPages(), [totalCount]);
+  const pages = useSelector((state) => state.todos.pages);
+  const page = useSelector((state) => state.status.page);
```
Comment (nortoza) [inline src/components/Pagination.js]: Lets call it currentPage

File: src/reducers/TodoReducer.js [hunk 19]
```diff
@@ -0,0 +1,60 @@
+import { GET_TODOS, ADD_TODO, UPDATE_TODO, UPDATE_TODOS } from "../constants";
+import { LIMIT } from "../constants";
```
Comment (nortoza) [inline src/reducers/TodoReducer.js]: Duplicate

File: src/reducers/TodoReducer.js [hunk 20]
```diff
@@ -0,0 +1,60 @@
+import { GET_TODOS, ADD_TODO, UPDATE_TODO, UPDATE_TODOS } from "../constants";
```
Comment (slimShady88) [inline src/reducers/TodoReducer.js]: Duplicate import

File: src/reducers/TodoReducer.js [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+import { LIMIT } from "../constants";
+
+const defaultState = {
```
Comment (slimShady88) [inline src/reducers/TodoReducer.js]: initialState is best practice

File: src/reducers/setStatusReducer.js [hunk 22]
```diff
@@ -0,0 +1,53 @@
+import {
+  SET_COMPLETED,
+  SET_PAGE,
+  SET_OFFSET,
+  SET_COMPLETED_ALL,
+  SET_COUNT,
+  SET_ERROR,
+  SET_FILTER,
+} from "../constants";
+import { LIMIT } from "../constants";
```
Comment (slimShady88) [inline src/reducers/setStatusReducer.js]: Duplicate import

File: src/reducers/setStatusReducer.js [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+
+const defaultState = {
```
Comment (slimShady88) [inline src/reducers/setStatusReducer.js]: The same

File: src/asyncActions/fetchTodos.js [hunk 24]
```diff
@@ -1,27 +1,26 @@
...
+      const response = await getTodos(TODOS_AT_PAGE, offset, completed);
       dispatch(getTodosAction(response));
-      dispatch(
-        setCompletedAllAction(
-          items.length === items.filter((item) => item.completed).length
-        )
-      );
-      dispatch(setCountAction(items.filter((item) => !item.completed).length));
+
+      const { todos } = response.todos;
```
Comment (slimShady88) [inline src/asyncActions/fetchTodos.js]: Why here .todos? If it's really needed: `const {todos: todos} = response`, but looks like useless logic

File: src/asyncActions/addTodo.js [hunk 25]
```diff
@@ -5,15 +5,15 @@ import {
   setCompletedAllAction,
```
Comment (slimShady88) [inline src/asyncActions/addTodo.js]: Lets move asyncActions to one file "asyncActions.js" - asyncActions still not moved to one file

File: src/actions/index.js [hunk 26]
```diff
@@ -1,47 +1,47 @@
...
-  UPDATE_TODOS,
-  SET_COMPLETED,
-  SET_PAGE,
-  SET_OFFSET,
-  SET_COMPLETED_ALL,
-  SET_COUNT,
-  SET_ERROR,
-  SET_FILTER,
-} from "../constants";
+import constants from "../constants";
```
Comment (slimShady88) [inline src/actions/index.js]: we can use destructurizing `const {...} = constants` to get all fields

File: src/constants/index.js [hunk 27]
```diff
@@ -1,15 +1,17 @@
...
-export const SET_COMPLETED = "SET_COMPLETED";
-export const SET_COMPLETED_ALL = "SET_COMPLETED_ALL";
-export const SET_PAGE = "SET_PAGE";
-export const SET_OFFSET = "SET_OFFSET";
-export const SET_COUNT = "SET_COUNT";
-export const SET_ERROR = "SET_ERROR";
-export const SET_FILTER = "SET_FILTER";
-
-export const LIMIT = 5;
+export default constants;
```
Comment (slimShady88) [inline src/constants/index.js]: We can write simply `export default const constants = {.....}`

File: src/asyncActions/index.js [hunk 28]
```diff
@@ -0,0 +1,117 @@
...
+  addTodo as appendTodo,
+  updateCompleted,
+  deleteTodo as removeTodo,
+  deleteCompleted,
+  getTodos,
+  updateTodo as changeTodo,
+} from "../api/todoApi";
+import constants from "../constants";
+
+const { TODOS_AT_PAGE } = constants;
```
Comment (nortoza) [inline src/asyncActions/index.js]: You can import this like "import { TODOS_AT_PAGE } from ..."

File: src/constants/index.js [hunk 29]
```diff
@@ -0,0 +1,17 @@
...
+  SET_COMPLETED_ALL: "SET_COMPLETED_ALL",
+  SET_PAGE: "SET_PAGE",
+  SET_OFFSET: "SET_OFFSET",
+  SET_COUNT: "SET_COUNT",
+  SET_ERROR: "SET_ERROR",
+  SET_FILTER: "SET_FILTER",
+  TODOS_AT_PAGE: 5,
+};
+
+export default constants;
```
Comment (nortoza) [inline src/constants/index.js]: Lets export it not as default

File: src/constants/index.js [hunk 30]
```diff
@@ -0,0 +1,17 @@
+const constants = {
```
Comment (nortoza) [inline src/constants/index.js]: Call it more specified in  CamelCase

File: src/reducers/TodoReducer.js [hunk 31]
```diff
@@ -0,0 +1,62 @@
+import constants from "../constants";
+
+const initialState = {
+  todos: [],
+  totalCount: null,
+  pagesCount: null,
+};
+
+const { GET_TODOS, ADD_TODO, UPDATE_TODO, UPDATE_TODOS, TODOS_AT_PAGE } =
```
Comment (nortoza) [inline src/reducers/TodoReducer.js]: I think we can dont use destruct and use it directly - CamelCaseName.GET_TODOS, etc..

File: src/reducers/setStatusReducer.js [hunk 32]
```diff
@@ -0,0 +1,55 @@
...
+  completed: null,
+  completedAll: false,
+  filter: "all",
+  offset: 0,
+  currentPage: 1,
+  count: 0,
+  errorMessage: "",
+};
+
+const {
```
Comment (nortoza) [inline src/reducers/setStatusReducer.js]: The same

File: src/config/config.js [hunk 33]
```diff
@@ -1,3 +1,4 @@
 export const config = {
   API_URL: process.env.REACT_APP_API_URL || "localhost",
+  TODOS_AT_PAGE: process.env.REACT_APP_TODOS_AT_PAGES || 5,
```
Comment (nortoza) [inline src/config/config.js]: Lets call it TODO_PER_PAGE

**Pull request merged**